scripts/22-starting-rust.md

```diff
@@ -0,0 +1,582 @@
+
+in other languages the compiler can't help you very much, so moving fast and breaking things is the best way develop.
+(any syntax gets in the way)
+
+In rust the compiler is your best friend.
+But you must be nice to it for it to be nice to you.
+
+
+---
+
+
+![[22-starting-rust 2023-03-20 20.38.27.excalidraw]]
+notes:
+
+Here's how I think the difficulty in learning, say Javascript and Rust compares to each other.
+I've picked Javascript because I want to make as many friends as possible, but the general relationship is the same for Python and Java, and many others too
+
+Rust up-fronts all the work that you would need to do in other languages later on to get scalability and performance and safety.
+
+And in making you deal with them at the start, the language debt doesn't spiral as you increase complexity.
+
+---
+
+> In other languages simple things are easy and complex things are possible, in Rust simple things are possible and complex things are EASY.
+
+---
+
+![[22-starting-rust 2023-03-20 20.49.45.excalidraw]]
+
+notes:
+And this is as it should be.
+I don't care how easy it is to do simple things, that doesn't impress me any more.
+I want a language that can stay by my side as I make extremely complex distributed systems that run at C speed and don't wake me up at night.
+
+Keep in mind through all your early learning, YOU only have to learn Rust once.
+Javascript developers have to keep learning javascript every single day.
+
+Edge case after edge case after edge case.
+Framework after framework to try and manage these edge cases.
+
+It's exhausting.
+Rust offers another way. But you have to work for it.
+
+For example
+
+---
+
+# Lifetime annotations
+
+```rust
+fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
+    if x.len() > y.len() {
+        x
+    } else {
+        y
+    }
+}
+```
+
+notes:
+
+Lifetime annotations specify how long a reference should live.
+It's a MINIMUM value that the variable must live for, before the borrow checker knows it is safe to clean up.
+
+This example is from Chapter 10 of The Book, which states:
+
+> the returned reference will be valid as long as both the parameters are valid. 
+> We'll name the lifetime `'a` and then add it to each reference
+
+It's very unfamiliar stuff, right?
+
+There's two ways to think about Lifetimes 
+1. As a liability and 
+2. an asset.
+
+I started off thinking about them as a liability - something annoying I HAD to do to satisfy the compiler. This is a totally fine way to start off with, and I have some suggestions on how to learn lifetimes gradually.
+
+In order
+
+---
+
+## 1. Don't use references
+## 2. Copy & Clone everything
+## 3. Obey the compiler
+
+notes:
+
+1. Don't use references, use ownership and share nothing. Pass in variables into functions, and if you need them later, return them back out, giving away ownership.
+
+2. Copy and clone everything. Rust is 80x faster than python, you're probably not going to notice the performance hit!
```

Comment:
I think it would be great to mention the [Clone to satisfy the borrow checker](https://rust-unofficial.github.io/patterns/anti_patterns/borrow_clone.html#clone-to-satisfy-the-borrow-checker) anti-pattern.

The best way to teach it for beginners I think is to say this:
> - using .clone() is like writing `let bar = foo;`. It basically copies the value and doesn't synchronize the data between the two.
> - If you want to syncronize it, you can use [Arc](https://doc.rust-lang.org/std/sync/struct.Arc.html) or [Rc](https://doc.rust-lang.org/std/rc/struct.Rc.html), though Rc can't be used in multi-threaded environments.
> - You can also look into [RefCell](https://doc.rust-lang.org/std/cell/struct.RefCell.html). It's a way of making the borrow rules at runtime rather than compile time. Just, remember that there's can only be one mutable access to it at one time, or the unwrap() will fail.
> Even though .clone() is an indication of a bad pattern, sometimes it is fine to write inefficient code, in cases such as when:
> 
> -    the developer is still new to ownership
> -    the code doesn't have great speed or memory constraints (like hackathon projects > - or prototypes)
> -  satisfying the borrow checker is really complicated, and you prefer to optimize readability over performance